services/cards/src/actions/Cards.tsx

```diff
@@ -8,10 +8,51 @@ import {filtersCalculate} from './Filters';
 declare var require: any;
 const Tabletop = require('tabletop');
 
+interface ResultType {
+  cards: CardType[];
+  translations: TranslationsType;
+}
+
 export function downloadCards(): ((dispatch: Redux.Dispatch<any>) => void) {
   return (dispatch: Redux.Dispatch<any>) => {
     const store = getStore();
     dispatch(cardsLoading());
+
+    const keys = store.getState().filters.source.current.split(':')[1].split(',');
+
+    Promise.all(keys.map(downloadAndProcessSpreadsheet))
+      .then((results: any[]) => {
+        const cards = results.reduce((acc: CardType[], obj: ResultType) => {
+          return [...acc, ...obj.cards];
+        }, []).sort((a: CardType, b: CardType) => {
+          if (a.sheet < b.sheet) {
+            return -1;
+          } else if (a.sheet > b.sheet) {
+            return 1;
+          } else if (a.class < b.class) {
+            return -1;
+          } else if (a.class > b.class) {
+            return 1;
+          }
+          return 0;
+        });
+
+        const translations = results.reduce((acc: TranslationsType, obj: ResultType) => {
+          return {...acc, ...obj.translations};
+        }, {});
+
+        if (Object.keys(translations).length > 0) {
+          dispatch(translationsUpdate(translations));
+        }
+        dispatch(cardsUpdate(cards));
+        dispatch(cardsFilter(store.getState().cards.data, store.getState().filters));
```

Comment:
This is also a lot of unnecessary `getState` for an action. Can we pass these in as arguments?

Reply:
So, this is how it was done previously, since cardsUpdate() actually causes data mutations that must be read into cardsFilter. Agree that it's not the greatest, but redoing the card processing / filtering system is out of scope here :(

Reply:
Woah, did not realize that -  it's not indicated by this code in any way and I don't think we have any other redux action that intentionally rely on their own side effects like this. This is a regression waiting to happen, since we also have no implemented unit tests for this code whatsoever.

If you look at what `cardsUpdate` does... it just sets `state.cards.data` to `cards` and doesn't even touch `state.filters`. So we really don't need to re-fetch the args for `cardsFilter` at all. Can you fix that?

`filtersCalculate()` is a bit trickier, so I'm OK with a big fat TODO comment indicating why we're getting state like this... provided we actually unit test this function and open a bug to pull the logic out of the reducer and into an action function.